Fix legacy BIOS boot on stricter BIOSes requiring an active MBR/DOS partition

digitalrane opened the pull request:
**Short version**: This MR fixes booting Chimera ISO images written to a USB device on older BIOS machines which require a bootable MBR/DOS style partition to exist on the USB device before showing it as a bootable device.

**Long version**: I was trying to install Chimera Linux on my corebooted t440p, and an older Sony VAIO laptop, and couldn't get the install media to recognize as a bootable device when written to a USB drive.

After experimenting for a little bit I noticed that the main difference between MBR/BIOS boot images which worked on these machines and those which didn't (like the Chimera Linux ISO images written to a USB device) was that the working ones used MBR/DOS partition tables with an "active" or "boot" flagged partition pointing a the boot partition (partition flag 0x80).

Tools like `isolinux` from the `syslinux` suite add this automatically, so these typically work well on older machines. Chimera live images use `grub2-mkrescue` (which is great, and super flexible, and compiles nicely under Chimera's toolchain). But the resulting images do not have a bootable MBR/DOS partition so they are not bootable on these types of BIOS-boot machines.

However, `xorriso` does have a flag (`--mbr-force-bootable`) for adding a compatible bootable partition to the protective/fake MBR DOS partition table embedded in the GPT partition table generated for hybrid boot images (BIOS+UEFI+CD). This flag produces images which are bootable by BIOS machines which previously did not boot the image, such as SeaBIOS and other machines I tested.

So, this MR enables a "fake" bootable partition in the protective MBR which is embedded in the GPT filesystem created by `xorriso`/`grub-mkrescue`. This allows stricter legacy BIOS implementations to detect the Chimera boot images as bootable when written to USB, and subsequently allows them to be booted and Chimera Linux installed on these machines.

Commit:
set gpt protective mbr boot partition as active

q66 commented:
i do wonder if there is any reason why grub-mkrescue (which is what we are using) does not do it by default? these images are definitely meant to be bootable on bios machines...

q66 commented:
i suppose this really only affects usb because optical media boot does not use mbr at all, but still

digitalrane commented:
There is an upstream discussion on how to handle this in the general case [here](https://savannah.gnu.org/bugs/?func=detailitem&item_id=46716), which has some context around why `grub-mkrescue` nor `xorriso` do this by default. My understanding of the whole situation is that in `xorriso`'s case it will set the appropriate flags if there are MBR partitions present, however `grub-mkrescue` only installs a protective MBR with no partitions, which is why the flag is needed to override the default behavior.

Other Linux distributions such as Arch have also documented this and included it in their scripts to make sure the partition is added with the bootable flag set. For example, [mkarchiso](https://gitlab.archlinux.org/archlinux/archiso/-/blob/master/archiso/mkarchiso#L1517). Not that I am suggesting that means Chimera Linux should do it just because another distribution does it, I add this only as additional context.

q66 commented:
i'm mostly curious about it because i'd like to avoid potentially breaking some other setup where things currently work (the hybrid boot stuff is a little messy)

at very least the flag should be added in a path that is only run for x86 images

q66 commented:
i followed the discussions a bit and they say it breaks booting from removable device on tianocore-based EFI systems... (unless that's a different case) so there may be others

i'll need to think about it and test a bunch of things

digitalrane commented:
In the grub bug report, the behavior/fix being described is different to what the `xorriso` flag does. The grub bug discussion talks about marking the protective MBR partition (1st partition in the protective MBR's embedded DOS partition table) as bootable, which is incorrect (and doesn't fix BIOS booting in my testing) and violates UEFI spec, causing issues on some stricter UEFI implementations. The `xorisso` flag instead adds a second partition pointing at the bootloader in the protective MBR's partition table and marks that bootable. So, the behavior is specifically designed to avoid compatibility issues with EFI-based systems, and you end up with the actual protective MBR partition untouched. So the risk would be limited to an UEFI implementation also validating that there is only a single partition in the protected MBR.

For non-x86 platforms, I think it makes sense to only set this on x86, but it also shouldn't impact anything, as my understanding is that they all talk GPT and look for data in the GPT partition table instead. I could be wrong as I don't have any powerpc gear here to test. I do have some UEFI-enabled uboot equipment I can test against, though.

q66 commented:
i think i might switch the live environment off grub entirely (limine supports all of the environments we target except powerpc and i could probably just keep that on grub but target powerpc-ieee1275 only)

q66 commented:
i added support for limine instead of grub in c502d31ed8bc0523e840b94b713e2acac12a9aa6 , would you mind trying generating an image with it (set `MKLIVE_BOOTLOADER=limine` in environment) and try it on your devices? it would be super helpful (i can try it on aarch64 and other efi places, i already tried it on ovmf efi and bios in x86 vm)

digitalrane commented:
No problem, I'll test it out. Looking at the limine code, it seems like it should work though, which is great! It looks like `limine` does the same thing as the `--mbr-force-bootable` flag in `xorriso`, but by default (see [limine.c](https://github.com/limine-bootloader/limine/blob/v8.x/host/limine.c#L861)). I'll let you know how my testing goes, thanks!

digitalrane commented:
Ah, that's actually on the codepath where limine is being run against an MBR/DOS partitioned disk. So I think the same flag is needed to make this work on older BIOS systems when using a GPT partitioning scheme (as is needed by UEFI). I tested the generated media and it worked on SeaBIOS, but not the VAIO I have here. Looking at the code again more closely, and the `xorriso` command suggested by `limine`, I think the same flag will be needed to enable stricter BIOS-boot machines. I'll test with the `--mbr-force-bootable` flag and and report back. Perhaps `limine` would be open to adding this functionality upstream.

digitalrane commented:
I did some more testing and I can't get the `limine` images to boot on the test VAIO I have (pre-UEFI x86_64) machine, even with the `--mbr-force-bootable` flag to `xorriso`. Oddly, SeaBIOS does work, but it's much newer and likely it works because it has support for GPT partition tables. It seems that `limine` is not putting the MBR part of the bootloader in a place that works reliably with the `--mbr-force-bootable` flag, as far as I can tell. For now, if you'd prefer to move toward `limine`, we can close this MR as I don't have a way to adapt these changes to `limine` in a reliable way. I'll continue to poke at `limine` itself though to see if I can get it working on the VAIO I'm testing with. Thanks for digging into this with me!

q66 commented:
it would be great if you could help me figure out a configuration that works, because i don't have any relevant hardware unfortunately (i may be able to try some in december, but not sure), we're not in a hurry but i'd like to figure it out before the next set of images is out...

q66 commented:
i did some testing with the argument in this PR on what i do have access to for now, and it seems to be ok, so i incorporated it, but i'd like to move off grub in the near future

q66 commented:
@digitalrane can you pull latest changes and try generating and running again? i found that limine was actually refusing to run on bios under certain circumstances (probably a bug) and was able to reproduce it in qemu with some thirdparty bios images; i have fixed that though and was able to have the bootloader come up when booting from both disk and cdrom

i tested a variety of configs (several bioses, tianocore uefi, ...) and they all work now, but i need it tested on real hardware...

digitalrane commented:
> @digitalrane can you pull latest changes and try generating and running again? i found that limine was actually refusing to run on bios under certain circumstances (probably a bug) and was able to reproduce it in qemu with some thirdparty bios images; i have fixed that though and was able to have the bootloader come up when booting from both disk and cdrom
> 
> i tested a variety of configs (several bioses, tianocore uefi, ...) and they all work now, but i need it tested on real hardware...

@q66 - happy to test this out. i pulled down the latest changes, built the image, flashed to a USB and unfortunately it's not detected as bootable by either my VAIO (non-EFI stock BIOS) or by SeaBIOS (as a coreboot payload).

i've done a bunch of tinkering with limine over the last couple of days and i've got a few clues as to why it's not working specifically around the protective MBR is creates in hybrid/ISO/UEFI mode.

i was able to get Limine to boot on the VAIO from a USB key and successfully boot Chimera Linux by putting together a test script which created a straight MBR partition table in an image file, formatted the single partition as VFAT, marked it bootable and then did a bios-install on the resulting disk image. but strangely, that wasn't picked up by SeaBIOS.

so I currently suspect the MBR/BIOS boot code in Limine is installing the stage1/2 loaders in a way they're not being seen as bootable, i checked and the MBR partition signatures and in the right place, and even when i manually add a second partition to the protective MBR which is bootable (similar to what the GRUB flag we added does) it still isn't detected, so i think more tinkering is needed here. i've been hex-dumping and comparing the MBR installed by this, isolinux and grub and haven't found anything yet which _should_ be causing an issue, but something must be off because i can boot both grub and isolinux hybrid images on these machines just fine

were you able to get the hybrid Limine images booting in qemu with seabios or ovmf or both? the seabios in coreboot should be very similar to the seabios in QEMU

q66 commented:
i was able to get them working with seabios, award bios, and ovmf, in either `-boot d -cdrom ...` or `-boot c -drive file=...` mode

q66 commented:
can you try these? https://ftp.octaforge.org/q66/random/chimera-linux-x86_64-LIVE-20241119-base.iso , https://ftp.octaforge.org/q66/random/test.iso

these both worked for me, the only difference is that test.iso contains hfsplus while the other one does not

q66 commented:
you can also try this one https://ftp.octaforge.org/q66/random/test2.iso

that one has a bootable partition in the protective mbr, otherwise is the same (also boots in all 6 variants for me)

digitalrane commented:
to follow up on what i mentioned on IRC, those images didn't make a difference on the machines I was testing, i'd tried various combinations of secondary MBR partitions etc. and that didn't make a difference previously either.

also i wanted to just clarify that the USB devices don't show up in the boot list, and that's the problem I'm seeing. with qemu etc. even with the USB device emulation afaict you're also telling qemu to boot from that device (it's in the boot order iirc) - i think there's some additional checks that are happening when determining what to add to the boot menu in SeaBIOS and those checks seem to be the main issue on SeaBIOS on the machines I'm testing on.

on the VAIO stock bios, it doesn't have a boot menu, and pretty much if there is a valid MBR with either no partition table or a partition table it is happy with, it will boot pretty reliably.

i spent some time playing around with different MBR and did manage to make a simple assembly MBR which both machines were happy with, and that just had 0xaa55 in the right signature location and a simple mbr which booted at offset 0x7c00, which is also what limine uses.

i noticed that limine kind of unusually embeds a PBP section in the MBR too, which is normally only for floppy disks, which might be confusing things. I removed this, and it did seem to be make a difference on the VAIO, but wasn't reliable, as I think there is some extra validation of bootable partitions happening or something there too.

what i did try which has been 100% reliable on every BIOS system I've tested, is just generating a simple MBR partition table with a single partition marked as active, formatted with FAT32, which contains the required files to boot Chimera Linux via Limine. this has worked 100% of the time for me but is not a hybrid image, so no el torito or EFI is present, so also no GPT. if you want to see what i mean, you can check out [this branch](https://github.com/digitalrane/chimera-live/tree/add-mbr-only-image) and build a simple disk image with `MKLIVE_BOOTLOADER=limine MKLIVE_IMAGE_TYPE=mbr ./mklive.sh`. it's a bit messy because it was a hack for testing but if you had any interest in publishing a separate "legacy" image which was just an MBR bootloader with a vfat partition we could clean something like this up and publish a second image type to avoid digging into limine much further.

i also spent some time digging around the assembly source for the MBR in both limine and grub and one big different is that grub doesn't relocate to 0x7c00. i also found that if i changed the relocation address to 0x8000 in limine it stopped being detected as bootable so i wonder if some kind of validation of the relocation address is happening too in some  bioses, but it could just be a total coincidence, it was hardly a scientific test.

q66 commented:
i don't really like the idea of a separate image, this issue should be fixable, we just need to figure it out
